### PR TITLE
Region Proposal Network (RPN) classification and regression losses (WIP)

### PR DESCRIPTION
**Pull request by 0x00b1**
Implementation of the Region Proposal Network (RPN) classification and regression losses.

Noteworthy items:

* We’ve refactored anchor generation as Keras backend functions so they’ll run from the Keras backend libraries like TensorFlow and Theano.

* The loss is computed on a Lambda layer. This is a workaround for Keras’ assumption that the `y_true` and `y_pred` values have identical shapes. We should probably refactor this into a custom layer to simplify model construction.

* It’s buggy. While there’re plentiful unit tests they likely don’t provide sufficient coverage. 

* We need to write documentation.

### CODE REVIEW

**Comment by JihongJu** [comment]
> Keras’ assumption that the y_true and y_pred values have identical shapes. 

That is interesting. I wasn't aware of it. Do you think we could add pre-processing to `targets` and generate `y_true` on the fly? 
[Keras model.compile L833](https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L833)

File: tests/test_models.py [hunk 1]
```diff
@@ -1,10 +1,90 @@
...
+    y_pred = keras.layers.concatenate([classification, regression])
+
+    model = keras.models.Model(image, features)
+
+    loss = keras_rcnn.losses.rpn.proposal(9, (224, 224), 16)
+
+    def example(x):
+        y_true, y_pred = x
+
+        return loss(y_true, y_pred)
```
Comment (JihongJu) [inline tests/test_models.py]: I think `loss` here needs to be passed as an argument to the Lambda layer. Otherwise, it will cause problems while saving and loading the model. See [#5396](https://github.com/fchollet/keras/issues/5396).

(Not tested)
```python
rpn_proposal_loss = keras_rcnn.losses.rpn.proposal(9, (224, 224), 16)
def rpn_loss(x, loss):
    y_true, y_pred = x
    return loss(y_true, y_pred)
```
loss = keras.layers.Lambda(example, arguments={'loss': rpn_proposal_loss},)([y_true, y_pred])

Comment (jhung0) [inline tests/test_models.py]: Cool! Incorporating that

**Comment by jhung0** [comment]
The code to handle the different shapes was inspired by https://github.com/fchollet/keras/issues/4781 suggests using dummy targets

File: keras_rcnn/layers/object_detection/_object_proposal.py [hunk 2]
```diff
@@ -1,15 +1,13 @@
...
 
 
 class ObjectProposal(keras.engine.topology.Layer):
-    """
-    Transforms RPN outputs (per-anchor scores and bounding box regression estimates) into object proposals.
-    """
-
     def __init__(self, maximum_proposals=300, **kwargs):
-        self.output_dim = (None, None, 4)
+        self.output_dim = (None, maximum_proposals, 4)
```
Comment (jhung0) [inline keras_rcnn/layers/object_detection/_object_proposal.py]: Yes seems so. What name would be good? `output_shape`?

File: keras_rcnn/backend/common.py [hunk 3]
```diff
@@ -61,21 +70,21 @@ def _mkanchors(ws, hs, x_ctr, y_ctr):
...
-                                         y_ctr + 0.5 * (hs - 1)), axis=1)
     return anchors
 
 
 def _ratio_enum(anchor, ratios):
     """
     Enumerate a set of anchors for each aspect ratio wrt an anchor.
     """
-
+    # import pdb
```
Comment (jhung0) [inline keras_rcnn/backend/common.py]: Yep missed that one

**Comment by 0x00b1** [comment]
> I just noticed this repository yesterday and, although it is far from finished, it looks really well thought out. I'm hoping I can contribute actively to this project in the near future. I was working on my own port of py-faster-rcnn because I was unhappy with the existing ports for keras / tensorflow, but got stuck on the losses for the RPN network. I mainly believe the issue is with how Keras forces the loss to be computed in a certain way and I made an issue about it on the keras github page (fchollet/keras#7395). If you would like to contribute to that discussion, that'd be helpful. I'm curious how the RPN loss is currently handled in keras-rcnn, that part is a bit unclear to me still. I'm guessing it is some form of workaround, but which kind is it ? :p

Hi, @hgaiser! Let’s collaborate! It’s a tricky problem and we could use the help!

In this pull request, the approach:

* For y_pred, the two feature maps produced by the convolutional layers that return the regression (for the reference bounding boxes) and the classification (for the reference bounding box’s objectness) are concatenated into one matrix.
* For y_true, we generate the anchors as described by the Faster RCNN paper. It’s somewhat different than the py-faster-rcnn implementation by @rbgirshick as the anchors are generated on the GPU (as part of the loss computation).

**Comment by hgaiser** [comment]
Hi,

I would very much like to collaborate! I had been working on a keras port by myself and it was driving me insane ;)

I made a few changes to this PR already, I'd like to share that with you tomorrow, I'll do so in another PR. That is sort of my view on how I imagine it, I would like to hear what you had in mind and what you think of my approach. How shall we talk? I'm always online on Slack (kerasteam), maybe that could be a good channel.

File: keras_rcnn/layers/object_detection/_object_proposal.py (hunk 2 above)
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/_object_proposal.py]: Yeah I think so.

File: tests/test_models.py [hunk 4]
```diff
@@ -8,3 +14,84 @@ def test_resnet50_rcnn():
...
+def test_rpn():
+    options = {
+        "activation": "relu",
+        "kernel_size": (3, 3),
+        "padding": "same"
+    }
+
+    image = keras.layers.Input((224, 224, 3))
+
+    y_true = keras.layers.Input((None, 4), name="y_true")
```
Comment (0x00b1) [inline tests/test_models.py]: This is a problem. You’d need to add a bounding box input to the `predict` method because the model will expect two inputs (an image and corresponding bounding boxes). A possible implementation is creating a custom layer that acts like `keras.layers.Input` during training but produces reference bounding boxes during test.

File: tests/test_models.py [hunk 5]
```diff
@@ -8,3 +14,84 @@ def test_resnet50_rcnn():
...
+
+
+    features = keras.layers.Conv2D(64, **options)(image)
```
Comment (0x00b1) [inline tests/test_models.py]: I’d use a ResNet implementation from the keras-resnet package rather than VGG since we’ll consider this standard and match everything else.

File: tests/test_models.py [hunk 6]
```diff
@@ -8,3 +14,84 @@ def test_resnet50_rcnn():
...
+
+
+    rpn_loss = keras_rcnn.losses.rpn.proposal(9, (224, 224), 16)
+
+    def example(x, loss):
+        y_true, y_pred = x
+        loss_c, loss_r = loss(y_true, y_pred)
+        return loss_c + loss_r
+
+        # use the following for testing:
```
Comment (0x00b1) [inline tests/test_models.py]: You can remove these comments.

File: tests/test_models.py [hunk 7]
```diff
@@ -8,3 +14,84 @@ def test_resnet50_rcnn():
...
+
+    classification = keras.layers.Conv2D(9 * 1, (1, 1), activation="sigmoid")(features)
+    regression = keras.layers.Conv2D(9 * 4, (1, 1))(features)
+
+    y_pred = keras.layers.concatenate([classification, regression])
+
+
+    rpn_loss = keras_rcnn.losses.rpn.proposal(9, (224, 224), 16)
+
+    def example(x, loss):
```
Comment (0x00b1) [inline tests/test_models.py]: Use a more descriptive identifier.

File: tests/test_models.py [hunk 8]
```diff
@@ -8,3 +14,84 @@ def test_resnet50_rcnn():
...
+    y_true = numpy.array([
+        [1, 1, 100, 100],
+        [0, 0, 40, 50],
+        [50, 99, 100, 203],
+        [111, 5, 131, 34],
+        [4, 60, 30, 90]])
+
+    y_true = numpy.expand_dims(y_true, 0)
+
+    model.fit([a, y_true], [None])
```
Comment (0x00b1) [inline tests/test_models.py]: You should also test `model.predict` and `model.evaluate`.

**Comment by mcquin** [comment]
Remaining work: https://github.com/broadinstitute/keras-rcnn/pull/52#issuecomment-319687963

File: keras_rcnn/layers/losses.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+class Regression(keras.layers.Layer):
+    def __init__(self, anchors, **kwargs):
+        self.anchors = anchors
+
+        self.is_placeholder = True
+
+        super(Regression, self).__init__(**kwargs)
+
+    def _loss(self, y_true, y_pred):
```
Comment (hgaiser) [inline keras_rcnn/layers/losses.py]: Same here.

File: keras_rcnn/layers/losses.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+        return 1.0 * (a / b)
+
+    def call(self, inputs):
+        y_true, y_pred = inputs
+
+        loss = self._loss(y_true, y_pred)
+
+        self.add_loss(loss, inputs=inputs)
+
+        return y_pred
```
Comment (hgaiser) [inline keras_rcnn/layers/losses.py]: It's a bit semantic, but we could return `loss` instead of `y_pred` here. It is ignored, but semantically it makes more sense (and potentially necessary in the future, if `Keras` decides to change the way losses are handled).

File: keras_rcnn/layers/object_detection/_object_proposal.py [hunk 11]
```diff
@@ -1,25 +1,83 @@
+import keras.backend
 import keras.engine
 
 import keras_rcnn.backend
 
+# FIXME: remove global
+RPN_PRE_NMS_TOP_N = 12000
```
Comment (hgaiser) [inline keras_rcnn/layers/object_detection/_object_proposal.py]: Just a note, `12000` is only used during training, not testing.

File: keras_rcnn/losses/rpn.py [hunk 12]
```diff
@@ -1,21 +1,107 @@
-import keras
```
Comment (hgaiser) [inline keras_rcnn/losses/rpn.py]: This entire file can be removed?

File: keras_rcnn/models.py [hunk 13]
```diff
@@ -49,15 +48,17 @@ class ResNet50RCNN(RCNN):
...
 
     :return model: a functional model API for R-CNN.
     """
 
-    def __init__(self, inputs, classes, rois=300):
+    def __init__(self, inputs, classes, rois=300, blocks=[3, 4, 6]):
         # ResNet50 as encoder
-        encoder = keras_resnet.models.ResNet50
+        image, _, _ = inputs
+        features = keras_resnet.models.ResNet50(image, blocks=blocks, include_top=False).output
```
Comment (hgaiser) [inline keras_rcnn/models.py]: I was thinking about this one ... do we want to call the model as a function (ie. `keras_resnet.models.ResNet50(blocks=blocks, include_top=False)(image)`) or do we want to incorporate every layer in the RCNN model? The difference is that in the summary, you'd see all the layers (the way it is now) or you would see a 'layer' called ResNet50. Let me know what you think.

File: keras_rcnn/layers/losses.py [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(self, anchors, **kwargs):
+        self.anchors = anchors
+
+        self.is_placeholder = True
+
+        super(Classification, self).__init__(**kwargs)
+
+    def _loss(self, y_true, y_pred):
+        # Binary classification loss
+        x, y = y_pred[:, :, :, :], y_true[:, :, :, self.anchors:]
```
Comment (hgaiser) [inline keras_rcnn/layers/losses.py]: I believe `y_true` is shaped (None, None) (as in (batch_id, label_id)). Slicing it like `y_true[:, :, :, self.anchors:]` won't work (why is slicing necessary on `y_true` anyway?)

File: keras_rcnn/layers/losses.py [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+        a = y_true[:, :, :, :self.anchors] * keras.backend.binary_crossentropy(x, y)
```
Comment (hgaiser) [inline keras_rcnn/layers/losses.py]: I have two questions here, why use `binary_crossentropy`? And why multiply with `y_true`?
I'm thinking `binary_crossentropy` should be used when there is only one output, but we have two. One option is to use only one of the two outputs (output for foreground, probably) and use `binary_crossentropy` (the option you seem to implement here), the other option is to use `categorical_crossentropy` and convert the labels to one_hot vectors. Which has the preference? I don't know :) intuitively I'd guess `categorical_crossentropy`. I'd love to hear your thoughts about this.

File: keras_rcnn/layers/losses.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+        a = keras.backend.sum(a)
+
+        # Divided by anchor overlaps
+        b = keras.backend.epsilon() + y_true[:, :, :, :self.anchors]
+        b = keras.backend.sum(b)
+
+        return 1.0 * (a / b)
```
Comment (hgaiser) [inline keras_rcnn/layers/losses.py]: Why not return `keras.backend.mean(keras.backend.binary_crossentropy(x, y))` directly? Why divide by anchor overlaps?

File: keras_rcnn/layers/losses.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(self, anchors, **kwargs):
+        self.anchors = anchors
+
+        self.is_placeholder = True
+
+        super(Regression, self).__init__(**kwargs)
+
+    def _loss(self, y_true, y_pred):
+        # Robust L1 Loss
+        x = y_true[:, :, :, 4 * self.anchors:] - y_pred
```
Comment (hgaiser) [inline keras_rcnn/layers/losses.py]: Same as `Classification`, I believe it is sized `(None, None, 4)` (batch_id, box_id, box).

File: keras_rcnn/layers/losses.py (hunk 14 above)
Comment (0x00b1) [inline keras_rcnn/layers/losses.py]: I think the arguments are backward. This is `y_pred`.

**Comment by 0x00b1** [comment]
@hgaiser Good call. Fixed `compute_output_shape` for the loss layers.

**Comment by hgaiser** [comment]
Oh that attribute was already there, nice :)

File: keras_rcnn/layers/object_detection/_anchor_target.py [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+        # map up to original set of anchors
+        labels = keras_rcnn.backend.unmap(labels, total_anchors, inds_inside, fill=-1)
+        bbox_reg_targets = keras_rcnn.backend.unmap(bbox_reg_targets, total_anchors, inds_inside, fill=0)
+
+        # TODO: implement inside and outside weights
+        return [labels, bbox_reg_targets]
+
+    def compute_output_shape(self, input_shape):
+        return [(None, 1), (None, 4)]
+
```
Comment (JihongJu) [inline keras_rcnn/layers/object_detection/_anchor_target.py]: I think a `get_config()` method would be useful to save and load trained model.

File: keras_rcnn/backend/tensorflow_backend.py [hunk 19]
```diff
@@ -5,18 +5,20 @@
...
 def matmul(
-    a,
-    b,
-    transpose_a=False,
-    transpose_b=False,
-    adjoint_a=False,
-    adjoint_b=False,
-    a_is_sparse=False,
-    b_is_sparse=False
+        a,
```
Comment (hgaiser) [inline keras_rcnn/backend/tensorflow_backend.py]: Just my curiosity here, why 8 spaces?
